iox_query/src/exec/gapfill/algo.rs

```diff
@@ -0,0 +1,267 @@
+        // change.
+        let sort_columns = sort_expr
+            .iter()
+            .take(sort_expr.len() - 1)
+            .map(|se| se.evaluate_to_sort_column(&batch))
+            .collect::<Result<Vec<_>>>()?;
+
+        let range_iter = if !sort_columns.is_empty() {
+            Box::new(lexicographical_partition_ranges(&sort_columns)?)
+        } else {
+            // lexicographical_partition_ranges will refuse to work if there are no
+            // sort columns. All the rows are in one big partition for this case.
+            Box::new(vec![(0..batch.num_rows())].into_iter())
+                as Box<dyn Iterator<Item = Range<usize>> + Send>
+        };
+
+        // For each time the group columns (not counting time) change...
+        for Range { start, end } in range_iter {
+            // For now, produce one output batch for each output series.
+            // TODO: be able to put multiple series into one output batch
+            //   and also be able to split a series over multiple output batches.
+
+            let output_builder = OutputBuilder::try_new(
+                output_batch_size,
+                input_stream.schema(),
+                &sort_expr,
+                &aggr_expr,
+            )?;
+
+            let sm = SeriesMetadata::new(input_time_array, start, end);
+            let output_batch = output_builder
+                .build_output(
+                    &gap_fill_params,
+                    input_time_array,
+                    &group_arr,
+                    &aggr_arr,
+                    &[sm],
+                )
+                .record_output(&baseline_metrics)?;
+
+            tx.send(Ok(output_batch))
+                .await
+                .map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
+        }
+    }
+    Ok(())
+}
+
+/// The parameters to the gap filling operator.
+/// Eventually this will contain start and stop time parameters
+/// that are pulled from the WHERE clause of the query.
+pub(super) struct GapFillParams {
+    pub stride: i64,
+}
+
+fn get_gap_fill_params(
+    batch: &RecordBatch,
+    stride: &Arc<dyn PhysicalExpr>,
+) -> Result<GapFillParams> {
+    // This parameter validation is similar to the validation for DATE_BIN
+    // in DataFusion. Users are not likely to see these errors since
+    // probably DATE_BIN validation would fail first.
+    let nanos = match stride.evaluate(batch)? {
+        ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(v))) => {
+            let (days, ms) = IntervalDayTimeType::to_parts(v);
+            let nanos =
+                (Duration::days(days as i64) + Duration::milliseconds(ms as i64)).num_nanoseconds();
+            match nanos {
+                Some(v) => v,
+                _ => {
+                    return Err(DataFusionError::Execution(
+                        "DATE_BIN_GAPFILL stride argument is too large".to_string(),
+                    ))
+                }
+            }
```

Comment:
you might be able to write this something like


```suggestion
           nanos.ok_or_else(||DataFusionError::Execution(
                        "DATE_BIN_GAPFILL stride argument is too large".to_string(),
           ))?
```

Reply:
Someday I will have all these little helper methods internalized...

Reply:
lol it took me a good 6 months and I am not sure I still have them all internalized